File: src/main/java/aurocosh/divinefavor/common/entity/minions/behaviour/MinionBehaviourSkeleton.java
package aurocosh.divinefavor.common.entity.minions.behaviour;

import aurocosh.divinefavor.common.entity.ai.EntityAIFollowOwner;
import aurocosh.divinefavor.common.entity.ai.EntityAIMinionWait;
import aurocosh.divinefavor.common.entity.ai.EntityAIOwnerHurtByTarget;
import aurocosh.divinefavor.common.entity.ai.EntityAIOwnerOrderedToAttack;
import aurocosh.divinefavor.common.entity.minions.base.IMinion;
import aurocosh.divinefavor.common.entity.minions.base.MinionData;
import aurocosh.divinefavor.common.entity.minions.behaviour.base.MinionBehaviour;
import net.minecraft.entity.EntityLiving;
import net.minecraft.entity.ai.*;
import net.minecraft.entity.monster.AbstractSkeleton;
import net.minecraft.entity.monster.EntitySkeleton;
import net.minecraft.entity.passive.EntityWolf;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.init.Items;
import net.minecraft.item.ItemStack;
import net.minecraft.world.EnumDifficulty;
import net.minecraftforge.fml.relauncher.ReflectionHelper;

public class MinionBehaviourSkeleton<T extends EntitySkeleton & IMinion> extends MinionBehaviour<T> {
    private static final int AI_ARROW_ATTACK_INDEX = 1;
    private static final int AI_ATTACK_ON_COLLIDE = 2;

    private static final int AI_ATTACK_PRIORITY = 5;

    private final EntityAIAttackRangedBow<AbstractSkeleton> aiArrowAttack;
    private final EntityAIAttackMelee aiAttackOnCollide;

    public MinionBehaviourSkeleton(T minion) {
        this.aiArrowAttack = ReflectionHelper.getPrivateValue(AbstractSkeleton.class, minion, AI_ARROW_ATTACK_INDEX);
        this.aiAttackOnCollide = ReflectionHelper.getPrivateValue(AbstractSkeleton.class, minion, AI_ATTACK_ON_COLLIDE);
    }

    @Override
    public void apply(T minion, EntityAITasks tasks, EntityAITasks targetTasks)  {
        MinionData minionData = minion.getMinionData();

        tasks.addTask(0, new EntityAISwimming(minion));
        tasks.addTask(1, new EntityAIRestrictSun(minion));
        tasks.addTask(2, new EntityAIFleeSun(minion, 1.0D));
        tasks.addTask(2, new EntityAIAvoidEntity<>(minion, EntityWolf.class, 6.0F, 1.0D, 1.2D));

        tasks.addTask(3, new EntityAIFollowOwner<>(minion, 2.0D, 5.0F, 2.0F, true, minionData::isFollowing));
        tasks.addTask(4, new EntityAIMinionWait<>(minion, () -> minionData.isFollowing() || minionData.isWaiting()));

        // attack task here

        tasks.addTask(6, new EntityAIWander(minion, 1.0D));
        tasks.addTask(6, new EntityAIFollowOwner<>(minion, 2.0D, 5.0F, 2.0F, false, () -> true));
        tasks.addTask(7, new EntityAIWatchClosest(minion, EntityPlayer.class, 8.0F));
        tasks.addTask(7, new EntityAILookIdle(minion));
        tasks.addTask(8, new aurocosh.divinefavor.common.entity.ai.EntityAIBeg<>(minion, 8.0F, Items.CHICKEN));

        targetTasks.addTask(1, new EntityAIOwnerHurtByTarget<>(minion));
        targetTasks.addTask(2, new EntityAIOwnerOrderedToAttack<>(minion, minionData));
        targetTasks.addTask(3, new EntityAINearestAttackableTarget<>(minion, EntityLiving.class, 0, true, true, minionData::shouldAttack));
        targetTasks.addTask(4, new EntityAIHurtByTarget(minion, false));
    }

    public void applyCombatBehaviour(T minion, EntityAITasks tasks) {
        tasks.removeTask(aiAttackOnCollide);
        tasks.removeTask(aiArrowAttack);
        ItemStack itemStack = minion.getHeldItemMainhand();

        if (itemStack.getItem() == Items.BOW) {
            int attackCooldown = minion.world.getDifficulty() != EnumDifficulty.HARD ? 40 : 20;
            aiArrowAttack.setAttackCooldown(attackCooldown);
            tasks.addTask(AI_ATTACK_PRIORITY, aiArrowAttack);
        }
        else
            tasks.addTask(AI_ATTACK_PRIORITY, aiAttackOnCollide);
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/entity/minions/behaviour/MinionBehaviourSkeleton.kt
package aurocosh.divinefavor.common.entity.minions.behaviour

import aurocosh.divinefavor.common.entity.ai.EntityAIFollowOwner
import aurocosh.divinefavor.common.entity.ai.EntityAIMinionWait
import aurocosh.divinefavor.common.entity.ai.EntityAIOwnerHurtByTarget
import aurocosh.divinefavor.common.entity.ai.EntityAIOwnerOrderedToAttack
import aurocosh.divinefavor.common.entity.minions.base.IMinion
import aurocosh.divinefavor.common.entity.minions.behaviour.base.MinionBehaviour
import net.minecraft.entity.EntityLiving
import net.minecraft.entity.ai.*
import net.minecraft.entity.monster.AbstractSkeleton
import net.minecraft.entity.passive.EntityWolf
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.init.Items
import net.minecraft.world.EnumDifficulty
import net.minecraftforge.fml.relauncher.ReflectionHelper

class MinionBehaviourSkeleton<T>(minion: T) : MinionBehaviour<T>() where T : IMinion, T : AbstractSkeleton {

    private val aiArrowAttack: EntityAIAttackRangedBow<AbstractSkeleton>
    private val aiAttackOnCollide: EntityAIAttackMelee

    init {
        this.aiArrowAttack = ReflectionHelper.getPrivateValue(AbstractSkeleton::class.java, minion, AI_ARROW_ATTACK_INDEX)
        this.aiAttackOnCollide = ReflectionHelper.getPrivateValue(AbstractSkeleton::class.java, minion, AI_ATTACK_ON_COLLIDE)
    }

    override fun apply(minion: T, tasks: EntityAITasks, targetTasks: EntityAITasks) {
        val minionData = minion.minionData

        tasks.addTask(0, EntityAISwimming(minion))
        tasks.addTask(1, EntityAIRestrictSun(minion))
        tasks.addTask(2, EntityAIFleeSun(minion, 1.0))
        tasks.addTask(2, EntityAIAvoidEntity(minion, EntityWolf::class.java, 6.0f, 1.0, 1.2))

        tasks.addTask(3, EntityAIFollowOwner(minion, 2.0, 5.0f, 2.0f, true, { minionData.isFollowing }))
        tasks.addTask(4, EntityAIMinionWait(minion, { minionData.isFollowing || minionData.isWaiting }))

        // attack task here

        tasks.addTask(6, EntityAIWander(minion, 1.0))
        tasks.addTask(6, EntityAIFollowOwner(minion, 2.0, 5.0f, 2.0f, false, { true }))
        tasks.addTask(7, EntityAIWatchClosest(minion, EntityPlayer::class.java, 8.0f))
        tasks.addTask(7, EntityAILookIdle(minion))
        tasks.addTask(8, aurocosh.divinefavor.common.entity.ai.EntityAIBeg(minion, 8.0f, Items.CHICKEN))

        targetTasks.addTask(1, EntityAIOwnerHurtByTarget(minion))
        targetTasks.addTask(2, EntityAIOwnerOrderedToAttack(minion, minionData))
        targetTasks.addTask(3, EntityAINearestAttackableTarget(minion, EntityLiving::class.java, 0, true, true) { minionData.shouldAttack(it) })
        targetTasks.addTask(4, EntityAIHurtByTarget(minion, false))
    }

    fun applyCombatBehaviour(minion: T, tasks: EntityAITasks) {
        tasks.removeTask(aiAttackOnCollide)
        tasks.removeTask(aiArrowAttack)
        val itemStack = minion.heldItemMainhand

        if (itemStack.item === Items.BOW) {
            val attackCooldown = if (minion.world.difficulty != EnumDifficulty.HARD) 40 else 20
            aiArrowAttack.setAttackCooldown(attackCooldown)
            tasks.addTask(AI_ATTACK_PRIORITY, aiArrowAttack)
        } else
            tasks.addTask(AI_ATTACK_PRIORITY, aiAttackOnCollide)
    }

    companion object {
        private val AI_ARROW_ATTACK_INDEX = 1
        private val AI_ATTACK_ON_COLLIDE = 2

        private val AI_ATTACK_PRIORITY = 5
    }
}
